### PR TITLE
Use the new GradleEnterpriseListener interface

### PR DESCRIPTION
**Pull request by facewindu**
This PR changes quite a bit the structure of the CCUDME project.
It is now an EventSpy that register very early (in its init callback) a custom listener implementing the new `GradleEnterpriseListener` interface. This allows us to get rid of Maven callbacks quirks and totally control what's going on here (instead of e.g. relying on AbstractMavenLifecycleParticipant callbacks being fired 'at the right time')

### CODE REVIEW

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/ApiAccessor.java
Comment (guylabs) [inline common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/ApiAccessor.java:12]: Are we able to provide this via the Plexus container from the Maven extension itself such that this code is handled on our side and then injected into the custom event spy?

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataMavenExtensionEventSpy.java
Comment (guylabs) [inline common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataMavenExtensionEventSpy.java:11]: Is this event spy required ? Can't we call `ApiAccessor.registerExtension(getClass(), container, logger);` inside the constructor of the `CommonCustomUserDataGradleEnterpriseListener` where we do have access to the `PlexusContainer` as well?

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/ApiAccessor.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/ApiAccessor.java:12]: no, because this code is especially about sharing things from the Maven extension to other places.

The only way to do that would be to hack parent realms from the maven extension side.
Basically, store in 'plexus.core' or another realm that is visible to all extensions in all cases the things we want to be visible to all extensions.
That would be possible I guess (it's another version of Marc's hack) but really not what Maven design wants us to do. (Marc's hack is about [this bug](https://issues.apache.org/jira/browse/MNG-6906) but does not violate Maven's principles, because if we understood exported packages properly, it should have worked even without the hack in all cases)

Then, even if we were ok doing this, we would still have to do the other way around: make the `GradleEnterpriseListener` implementation on the consuming extension available on the Maven extension side:
(this guy:
```
importPackage(extensionRealm, maybeGERealm.get(), OWN_PACKAGE);
```
)

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataMavenExtensionEventSpy.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataMavenExtensionEventSpy.java:11]: There are cases where the `CommonCustomUserDataGradleEnterpriseListener` is not constructed, even if marked with the `@Component` annotation.
But even there, it needs the `GradleEnterpriseListener` interface at runtime when constructed, so the interface needs to be available in its realm when constructed.(and luckily, it's constructed after `EventSpy#init` is called)

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CustomBuildScanEnhancements.java
Comment (marcphilipp) [inline common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CustomBuildScanEnhancements.java:1]: Maybe a bit shorter?
```suggestion
package com.gradle.maven.custom.extension;
```

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java
Comment (marcphilipp) [inline common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java:18]: Is this how IDEA indents it?

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java
Comment (marcphilipp) [inline common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java:55]: Should we maybe handle this at the call site in the GE Maven extension?

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/GroovyScriptUserData.java
Comment (marcphilipp) [inline common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/GroovyScriptUserData.java:18]: Should we instead document `session` and `project` as nullable and evaluate it anyway?

File: common-custom-user-data-maven-extension/pom.xml
Comment (marcphilipp) [inline common-custom-user-data-maven-extension/pom.xml:68]: Will we have to wait for the 1.11 release before merging this PR?

**Comment by etiennestuder** [comment]
I want to review this PR before it gets merged.

**Comment by etiennestuder** [comment]
@facewindu why did the package names have to change?

**Comment by facewindu** [comment]
> I want to review this PR before it gets merged.

don't worry @etiennestuder you will. I wanted to review it between Guy, Marc and me before asking the solutions team. It will need the Maven extension 1.11 anyway, as a dependency.

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/ApiAccessor.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/ApiAccessor.java]: Spent many hours on this, and finally got something. Which made the consuming part (i.e. the CCUDME) much nicer and shorter.

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CustomBuildScanEnhancements.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CustomBuildScanEnhancements.java]: I went back with the existing package  `com.gradle`. We originally needed to change it, because I was importing it back to the GE extension realm, and having sth not precise enough was causing some recursion in class loading. This is not a problem anymore as we don't import anything back

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java]: for me, yes

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/GroovyScriptUserData.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/GroovyScriptUserData.java]: not sure what you mean. If the session is not available (very edge case scenarios with very early Maven build failures), we cannot grab the location of the groovy script (from `session.getRequest().getMultiModuleProjectDirectory()` +`.mvn/gradle-enterprise-custom-user-data.groovy`)

File: common-custom-user-data-maven-extension/pom.xml
Comment (facewindu) [reply common-custom-user-data-maven-extension/pom.xml]: yes, we will have to wait for BA Maven 1.11 to be out first. This version will ship the new `GradleEnterpriseListener` interface, which is used instead of hooking ourselves into Maven lifecycle directly (via `AbstractMavenLifecycleParticipant` or event spies)

**Comment by facewindu** [comment]
> @facewindu why did the package names have to change?

it's not needed anymore

**Comment by facewindu** [comment]
@guylabs @marcphilipp please review again

@gradle/ge-solutions please review this PR.
This builds on top of https://github.com/gradle/dotcom/pull/10070 which allows to hook directly into our own lifecycle, not relying on Maven's.
This has the benefit to really control _when_ the callback is called, and ensure it is really always called. (It was not in case of very early Maven failures otherwise)

We also don't need to hacks to import the GE Maven extension API classes into this extension's realm anymore. It's all done on the GE Maven extension side, by inspecting existing class realms (and realms added after the GE Maven extension is initialized).

Basically now, the CCUDME only provides a `@Component` which implements our own public `GradleEnterpriseListener` interface. Everything is much simple on the consuming side, and more reliable.

I bumped the CCUDME version to 1.8 (new minor) because of this rather impactful change.
It also requires GE Maven extension 1.11+ (no backward compatibility, but I understood this was not an issue)

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java]: yes, I think this is a better idea. If the others are ok, I will do this.
Basically `GradleEnterpriseListener#configure` will declare that it can throw a `MavenExecutionException`, and we'll catch it around each listener call on the GE Maven extension side.

File: common-custom-user-data-maven-extension/README.md
Comment (runningcode) [inline common-custom-user-data-maven-extension/README.md:27]: What type of error message do we get if the versions are incompatible? Will it simply not do anything and silently fail?

**Review by runningcode — APPROVED** [review]
Looks good to me. I'm no maven expert though. @bigdaz might have some thoughts.

File: common-custom-user-data-maven-extension/pom.xml
Comment (bigdaz) [inline common-custom-user-data-maven-extension/pom.xml:68]: Since this is used as a template for customers, we should avoid having unreleased versions merged into `master`. I guess we can keep this PR in a branch until `1.11` is released.

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/CommonCustomUserDataGradleEnterpriseListener.java
Comment (bigdaz) [inline common-custom-user-data-maven-extension/src/main/java/com/gradle/CommonCustomUserDataGradleEnterpriseListener.java:27]: Is there not a way to implement the caller of `configure` such that `MavenSession` is never null? 
Forcing the client to deal with this seems unfortunate.

**Review by bigdaz — APPROVED** [review]
I think this is a big improvement, making the CCUDME a first-class collaborator with the GE Maven extension.
But we should not merge this into `master` until GE Maven extension 1.11 is released. Otherwise customers will not be able to use these sources as a template for their own custom extension.

File: common-custom-user-data-maven-extension/pom.xml
Comment (facewindu) [reply common-custom-user-data-maven-extension/pom.xml]: yes, that was the plan. I'll change the PR to contain 1.11 when released and we will merge it after. Keeping it like that so that you can play around with the custom 1.11 for now

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/CommonCustomUserDataGradleEnterpriseListener.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/CommonCustomUserDataGradleEnterpriseListener.java]: I just made changes on https://github.com/gradle/dotcom/pull/10070 and here, as a consequence, to do just that. It's indeed better.

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/CommonCustomUserDataGradleEnterpriseListener.java]: done

File: common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/GroovyScriptUserData.java
Comment (facewindu) [reply common-custom-user-data-maven-extension/src/main/java/com/gradle/maven/common/user/data/extension/GroovyScriptUserData.java]: moot

File: common-custom-user-data-maven-extension/README.md
Comment (facewindu) [reply common-custom-user-data-maven-extension/README.md:27]: It's silently ignored. We do get a 
```
TRACE: Sisu - Ignoring Plexus role: com.gradle.maven.extension.api.GradleEnterpriseListener
java.lang.TypeNotPresentException: Type com.gradle.maven.extension.api.GradleEnterpriseListener not present
...
```
exception logged at the `trace` level (the deepest) but no one will see that.

Given the choice I made to _only_ provide a `@Component` in the extension, we cannot do more than that. We are sure the components will be registered on 'SessionStarted' for instance, so we could also provide a `AbstractMavenLifecycleParticipant` that checks if the `GradleEnterpriseListener` type is present on the CCUDME realm, and log sth if that's not the case, but I think this is taking a step back (complexifying back the CCUDME).

**Review by guylabs — APPROVED** [review]

**Comment by guylabs** [comment]
@etiennestuder can you review this as well such that we can finalize it once @facewindu is back next week? Thanks.

**Review by marcphilipp — APPROVED** [review]

**Comment by facewindu** [comment]
This is fully ready, including declaring the yet-to-be-released Maven extension 1.11 in the pom.
@etiennestuder please have a last look.
This PR cannot be merged before Maven extension 1.11 is released

**Review by etiennestuder — APPROVED** [review]
Wonderful!

**Pull request merged**
